jujugui/static/gui/src/app/extensions/app-renderer-extension.js

```diff
@@ -37,43 +37,38 @@ YUI.add('app-renderer-extension', function(Y) {
 
       @method _renderBreadcrumb
       @param {Object} options
-        showEnvSwitcher: false
+        showEnvSwitcher: true
     */
     _renderBreadcrumb: function({ showEnvSwitcher=true } = {}) {
-      // If this.env is undefined then do not render the switcher because there
-      // is no env to connect to. It will be undefined when the breadcrumb
-      // is rendered in the callback for generateSocketUrl because an env
-      // has not yet been created.
-      var env = this.env;
-      var jem = this.jem;
-      var utils = views.utils;
-      // If gisf is enabled then we won't be connected to a model to know
-      // what facades are supported but we can reliably assume it'll be Juju 2
-      // or higher which will support the necessary API calls.
-      if (!this.get('gisf')) {
-        if(!env || env.findFacadeVersion('ModelManager') === null &&
-           env.findFacadeVersion('EnvironmentManager') === null) {
-          // We do not want to show the model switcher if it isn't supported as
-          // it throws an error in the browser console and confuses the user
-          // as it's visible but not functional.
-          showEnvSwitcher = false;
-        }
+      const env = this.env;
+      const controller = this.controllerAPI;
+      const utils = views.utils;
+      const state = this.state;
+      let listModels =
+        controller && controller.listModelsWithInfo.bind(this.controllerAPI);
+      // If controller is undefined then do not render the switcher because
+      // there is no controller to connect to. It will be undefined when the
+      // breadcrumb is initially rendered because it hasn't yet been given
+      // time to connect and login.
+      if (!controller || (this.get('sandbox') && !this.get('gisf'))) {
```

Comment:
Perhaps all this logic should deserve a more readable external function? Like 
```javascript
_shouldDisplayModelSwitcher: function(controllerAPI) {
  // If controller is undefined then do not render the switcher because
  // there is no controller to connect to. It will be undefined when the
  // breadcrumb is initially rendered because it hasn't yet been given
  // time to connect and login.
  if (!controllerAPI) {
    return false;
  }
  // It's possible that we have a controller instance and no facade if
  // we've connected but have not yet successfully logged in. This will
  // prevent the model switcher from rendering but after the login this
  // component will be re-rendered.
  if (controllerAPI.findFacadeVersion('ModelManager') === null) {
    return false;
  }
  // Always show the model switcher when the GUI is embedded in
  // the storefront.
  if (this.get('gisf')) {
    return true;
  }
  // If we're in sandbox we don't want to display the switcher.
  if (this.get('sandbox')) {
    return false;
  }
  // At this point we are connected and logged in to a real Juju
  // controller: definitively show the model switcher.
  return true;
}
```